src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,153 @@
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        Object[] buffer = toArray(list);
+        Object[] elementData;
+        int requiredCapacity = list.size() + size;
+        if (requiredCapacity >= currentCapacity) {
+            grow(requiredCapacity);
+        }
+        System.arraycopy(buffer, 0, elementData = this.elementData, size, list.size());
+        size += list.size();
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isIndexOutOfSize(index)) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Requested index " + index + " is out of bounds");
+        } else {
+            return elementData[index];
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (isIndexOutOfSize(index)) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Requested index " + index + " is out of bounds");
+        } else {
+            elementData[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (isIndexOutOfSize(index)) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Requested index " + index + " is out of bounds");
+        } else {
+            final int newSize;
+            T oldValue = elementData[index];
+            if ((newSize = size - 1) > index) {
+                System.arraycopy(elementData, index + 1, elementData,
+                        index, newSize - index);
+            }
+            elementData[size = newSize] = null;
+            return oldValue;
+        }
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int newSize;
+        int index = indexOf(element);
+        if (index != NO_INDEX) {
+            if ((newSize = size - 1) > index) {
+                System.arraycopy(elementData, index + 1, elementData,
+                        index, newSize - index);
+            }
+            elementData[size = newSize] = null;
+        } else {
+            throw new NoSuchElementException(
+                    "Requested element " + element + " is not in the list");
+        }
+        return element;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void growIfArrayIsFull() {
+        if (size == currentCapacity) {
+            currentCapacity = (int) (currentCapacity * DEFAULT_LOAD_FACTOR);
+            increaseCapacity();
+        }
+    }
+
+    private void grow(int minCapacity) {
+        currentCapacity = (int) (minCapacity * DEFAULT_LOAD_FACTOR);
+        increaseCapacity();
+    }
+
+    private void increaseCapacity() {
+        Object[] newData = new Object[currentCapacity];
+        for (int i = 0; i < elementData.length; i++) {
```

Comment:
Use System.arraycopy() instead of a loop for copying array elements. It is more efficient and makes the code cleaner.